### DISCUSSION TITLE
Missing node in exergy analysis

### DISCUSSION THREAD

**Discussion by HaSchneider** [oemof/tespy#539] [Q&A] [answered]
Hi all!
I have a problem with the exergy analysis of a steam cycle with desuperheater injection. When I make an exergy analysis for my model, the heat sink and the corresponding product exergy flow is empty (at least in the sankey visualization). 

Any ideas what I did wrong?

Here is a minimal example of the problem:
```python
from tespy.networks import Network
from tespy.components import (
    Source, Sink, CycleCloser, SimpleHeatExchanger, Merge, Splitter,
)
from tespy.connections import Connection, Ref, Bus
from tespy.tools import ExergyAnalysis

#creation of components:
nw = Network()
nw.set_attr(iterinfo=False)
nw.set_attr(T_unit='C', p_unit='bar', h_unit='kJ / kg')
heat_sink = SimpleHeatExchanger('heat sink')
merge_injection = Merge("Injection")
dummy_sink= Sink('dummy sink')
injection_source =Source('injection_source')
condensate_split= Splitter("split condensate")
cycl=CycleCloser('CycleCloser')
boiler = SimpleHeatExchanger('steam boiler')
inj_cycle = CycleCloser('injection cycle')

# Setup of connections
superheated_steam= Connection(boiler, 'out1', merge_injection, 'in1')
inj_condensate = Connection(injection_source, 'out1', merge_injection, 'in2')
saturated_steam = Connection(merge_injection, 'out1', heat_sink, 'in1')
condensate_1 = Connection(heat_sink, 'out1', condensate_split, 'in1')
inj_condensate_return = Connection(condensate_split, 'out2', dummy_sink, 'in1')
condensate_return_1 = Connection(condensate_split, 'out1', cycl, 'in1')

condensate_return_2= Connection(cycl,'out1', boiler, 'in1' )

nw.add_conns(condensate_1, 
             inj_condensate_return, 
             inj_condensate,
             saturated_steam,
             superheated_steam,
             condensate_return_1, 
             condensate_return_2)

# specify attributes:
superheated_steam.set_attr(fluid={'water':1}, p=40, Td_bp= 10)
condensate_1.set_attr(x=0)
condensate_return_1.set_attr(m=Ref(superheated_steam,1,0))
inj_condensate.set_attr(x=0, fluid={'water':1})

saturated_steam.set_attr(x=1)
heat_sink.set_attr(pr=1)

#Add busses:
product_bus =Bus('heat bus', P=-10000)
product_bus.add_comps({'comp':heat_sink})

fuel_bus = Bus('boiler bus')
fuel_bus.add_comps({'comp':boiler, 'base':'bus'})

water_bus = Bus('water bus')
water_bus.add_comps({'comp':injection_source, 'base':'bus'})

waste_water_bus = Bus('wastewater bus')
waste_water_bus.add_comps({'comp':dummy_sink, 'base':'component'})

nw.add_busses(fuel_bus, product_bus, waste_water_bus, water_bus)

nw.solve('design')

ean = ExergyAnalysis(nw, 
                    E_P=[product_bus], 
                    E_F=[fuel_bus, water_bus],
                    E_L=[waste_water_bus]
                    )
ean.analyse(pamb=1, Tamb=20)

import plotly.graph_objects as go

links, nodes = ean.generate_plotly_sankey_input()

fig = go.Figure(go.Sankey(
    arrangement='snap',
    node={
        'label': nodes,
        'pad':11,
        'color': 'orange'
    },
    link=links
))
fig.show()
```
![grafik](https://github.com/user-attachments/assets/9bff5c7f-a995-486e-ab34-274028ab77e9)

![steam_net_tespy_problem](https://github.com/user-attachments/assets/053607b9-cd62-4b0b-b5ae-8fe1fcd93822)

**Comment by fwitte** [comment, marked as answer]
Hi @HaSchneider and thanks for reaching out!

The issue is in the `SimpleHeatExchanger` being defined as dissipative component by default. I should change it, so the user gets an error, that the component has to be specified as dissipative or not explicitly. You can fix this in the current version of tespy by changing this line:

```python
heat_sink = SimpleHeatExchanger('heat sink', dissipative=False)
```

A dissipative component has no defined exergy product. This would be the case, if the energy provided to the heat sink would just be considered a heat loss to the ambient.

![grafik](https://github.com/user-attachments/assets/c8ac46aa-4590-494c-b3d3-67eda9c8dcf0)

**Reply by fwitte** [reply]
I will include an explicit error message for cases, in which the user did not specify the dissipative attribute in the next version of tespy https://github.com/oemof/tespy/issues/540.

**Reply by HaSchneider** [reply]
Thanks for the fast answer, that solved the problem! 
I am still struggling with the definition of `base`. When should I choose what specification? 

The next problem I have: 
The network does not converges, if I add a cycle closer for the condensate injection recycle ("Singularity in jacobian matrix, ..."):
```
nw.del_conns(inj_condensate,inj_condensate_return)
inj_condensate = Connection(inj_cycle, 'out1', merge_injection, 'in2')
inj_condensate_return = Connection(condensate_split, 'out2', inj_cycle, 'in1')
nw.add_conns(inj_condensate,inj_condensate_return)
condensate_return_1.set_attr(m=None)
nw.del_busses(waste_water_bus, water_bus)
nw.solve('design')
```

**Reply by fwitte** [reply]
`"base": "component"` (default) means, that the efficiency is relative to the components energy input to the bus. The source of energy is the component.
`"base": "bus"` means, that the source of energy is at the bus and flowing into the component (e.g. a pump, compressor).

For sources and sinks (in case you have material flows as exergy fuel, product or in case of exergy loss) you have the same logic: A source gets its energy from the bus (therefore base = bus) a sink transfers energy to the bus (therefore base = component), see https://github.com/fwitte/SEGS_exergy/blob/8fc548ff863313d149b35efd5ec24d9e50c442f3/SEGS.py#L319

To the `CycleCloser` issue: The component actually enforces pressure and enthalpy equality on both sides, mass flow is decoupled. What you want to to is this:

```python
# condensate_return_1.set_attr(m=None) --> Not this!
heat_sink.set_attr(pr=None)
```

The heat sink pressure ratio now must be a result because the cycle closer enforces same pressure at condensate split and injection merge. Therefore the pressure ratio over the heat sink must be a result and cannot be specified. The mass flow specification is still required in your case.

**Reply by fwitte** [reply]
On a quick side node: You actually only need a single cycle closer. That must be between the merge and the splitter (downstream of merge and upstream of splitter). This also removes the necessity of specifying the mass flow equality along the boiler side.

```python
from tespy.networks import Network
from tespy.components import (
    Source, Sink, CycleCloser, SimpleHeatExchanger, Merge, Splitter,
)
from tespy.connections import Connection, Ref, Bus
from tespy.tools import ExergyAnalysis

#creation of components:
nw = Network()
nw.set_attr(iterinfo=False)
nw.set_attr(T_unit='C', p_unit='bar', h_unit='kJ / kg')
heat_sink = SimpleHeatExchanger('heat sink', dissipative=False)
merge_injection = Merge("Injection")
dummy_sink= Sink('dummy sink')
injection_source =Source('injection_source')
condensate_split= Splitter("split condensate")
cycl=CycleCloser('CycleCloser')
boiler = SimpleHeatExchanger('steam boiler')
inj_cycle = CycleCloser('injection cycle')

# Setup of connections
superheated_steam= Connection(boiler, 'out1', merge_injection, 'in1')
saturated_steam = Connection(merge_injection, 'out1', heat_sink, 'in1')
condensate_1 = Connection(heat_sink, 'out1', cycl, 'in1')
condensate_2 = Connection(cycl, 'out1', condensate_split, 'in1')
inj_condensate_return = Connection(condensate_split, 'out2', merge_injection, 'in2')
condensate_return_1 = Connection(condensate_split, 'out1', boiler, 'in1')

nw.add_conns(
    condensate_1,
    inj_condensate_return,
    condensate_2,
    saturated_steam,
    superheated_steam,
    condensate_return_1,
)

# specify attributes:
condensate_2.set_attr(fluid={'water':1}, p=40, x=0)
superheated_steam.set_attr(Td_bp=10)

saturated_steam.set_attr(x=1)

#Add busses:
product_bus =Bus('heat bus', P=-10000)
product_bus.add_comps({'comp': heat_sink, "base": "component"})

fuel_bus = Bus('boiler bus')
fuel_bus.add_comps({'comp':boiler, 'base':'bus'})

water_bus = Bus('water bus')
water_bus.add_comps({'comp': injection_source, 'base':'bus'})

waste_water_bus = Bus('wastewater bus')
waste_water_bus.add_comps({'comp': dummy_sink, 'base':'component'})

nw.add_busses(fuel_bus, product_bus, waste_water_bus, water_bus)

nw.solve('design')
```

**Reply by HaSchneider** [reply]
Ah, yes, now it works! And yes, good idea to use only one recycle, I will try that! 
Thank you!
